### PR TITLE
feat(n4s): add enforce.record for dynamic object validation

### PR DESCRIPTION
**Pull request by ealush**
Implements enforce.record for dynamic object key/value validation. Supports both value-only validation and key+value validation.

## Release Notes

* **New Features**
  * Added `enforce.record()` rule for validating dynamic key-value objects with support for validating values only or both keys and values simultaneously
  * Supports nesting within shapes and arrays with accurate error path reporting

* **Documentation**
  * Added comprehensive documentation for `enforce.record()` with usage examples

### CODE REVIEW

**Comment by ealush** [comment]
I have addressed the review feedback by fixing the benchmark name typography and updating the `RecordRuleInstance` generic types to extract the correctly inferred parsed key, and ensuring that parsed/coerced keys are written to the output object correctly.

File: packages/n4s/src/rules/schemaRules/record.ts [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+ * @param value - The object to validate
+ * @param arg1 - Either the key rule (if arg2 is present) or the value rule
+ * @param arg2 - The value rule (if arg1 is the key rule)
+ * @returns RuleRunReturn indicating success or failure
+ */
+export function record<T extends Record<string, any>>(
+  value: T,
+  arg1: any,
+  arg2?: any,
+): RuleRunReturn<T> {
```
[the remark this answers is not in the document]
Comment (ealush) [reply packages/n4s/src/rules/schemaRules/record.ts:27]: Already handled — the code stores coerced values in `parsedValue` via `parsedValue[parsedKey] = valRes.type` and returns `RuleRunReturn.Passing(parsedValue as T)`. This is the same pattern used by `shape`.

File: packages/n4s/src/rules/schemaRules/record.ts [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+  const fieldValue = value[key];
+  const valRes = ctx.run({ value: fieldValue, set: true, meta: { key } }, () =>
+    rules.valueRule.run(fieldValue),
+  );
+
+  if (!valRes.pass) return createRecordFailure(value, key, valRes);
+
+  parsedValue[key] = valRes.type;
+}
```
[the remark this answers is not in the document]
Comment (ealush) [reply packages/n4s/src/rules/schemaRules/record.ts:91]: Already handled in the current implementation. The code uses `parsedKey = keyRes.type`, writes to `parsedValue[parsedKey]`, and deletes the old key when it differs from the original:

```typescript
parsedKey = keyRes.type;
// ...
parsedValue[parsedKey] = valRes.type;
if (parsedKey !== key) {
  delete parsedValue[key];
}
```

File: packages/n4s/src/rules/schemaRules/record.ts [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+  return newRes as RuleRunReturn<T>;
+}
+
+export type RecordRuleInstance<
+  _K,
+  V extends RuleInstance<any, any>,
+> = RuleInstance<
+  Record<string, V['infer']>,
+  [Record<string, InferShapeInput<V>>]
+>;
```
[the remark this answers is not in the document]
Comment (ealush) [reply packages/n4s/src/rules/schemaRules/record.ts:126]: Fixed. The original inlined conditional was distributing over `never`, producing `Record<never, V>` (equivalent to `{}`) which accepted any input. Extracted into `RecordKey<K>` / `RecordInputKey<K>` helpers that use `[K] extends [never]` to properly default to `string` keys. See commit 850c78d.

File: benchmark-results.md [hunk 4]
```diff
@@ -2,67 +2,67 @@
...
+| Nested Fields with Hooks          | depth 4 with 60 fields per level   | **6.369**    | 180.58   | 32.42%          | 0          | 0.00%    |
+| Nested Fields with Hooks          | depth 5 with 80 fields per level   | **6.254**    | 161.94   | 16.12%          | 0          | 0.00%    |
+| Complex Feature Mix               | full run with feature flags        | **135.67**   | 12.2272  | 7.84%           | 0          | 0.00%    |
+| Complex Feature Mix               | focused/conditional run            | **234.4**    | 7.7559   | 2.69%           | 0          | 0.00%    |
+| Deep Nesting Stress               | depth 10                           | **78.512**   | 35.3061  | 7.05%           | 0          | 0.00%    |
+| Deep Nesting Stress               | depth 50                           | **32.109**   | 38.3066  | 2.14%           | 0          | 0.00%    |
+| Deep Nesting Stress               | depth 100                          | **20.985**   | 52.4618  | 1.41%           | 0          | 0.00%    |
+| Complex Combinations & Edge Cases | High Frequency test Creation       | **191.36**   | 8.738    | 2.68%           | 0          | 0.00%    |
```
[the remark this answers is not in the document]
Comment (ealush) [reply benchmark-results.md:59]: Addressed in commit ba7c715.

File: packages/vest/src/suite/__tests__/schema.runtime.test.ts [hunk 5]
```diff
@@ -173,7 +173,7 @@ describe('Schema Runtime Validation', () => {
 
       const testSuite = create(() => {}, schemaWithMessage);
 
-      const result = testSuite.run({ price: 'free' } as any);
+      const result = testSuite.run({ price: 'free' });
```
[the remark this answers is not in the document]
Comment (ealush) [reply packages/vest/src/suite/__tests__/schema.runtime.test.ts:176]: Not applicable. The `.message()` chain widens the type, so `{ price: 'free' }` does not produce a TS error in this case. Adding `@ts-expect-error` causes an "unused directive" error (TS2578). The other test without `.message()` correctly has the directive.
